docs/source/siteconfig.rst
.. _site_configuration:

Site Configuration
==================

The Mastr-MS site variables are defined in the settings.py file found within the application root. Many of these have sensible defaults, but some need to be populated in order for the site to work - database settings, email servers etc. This can be done in the settings.py file directly, although this is not advised, since redeploying the site or any updates to the source code will overwrite your custom settings. The best strategy is to leverage the final import statement in the settings.py file, which is used to pull in deployment specific overrides::

    # Override defaults with your local instance settings.
    # They will be loaded from appsettings.<projectname>, which can exist anywhere
    # in the instance's pythonpath. This is a CCG convention designed to support
    # global shared settings among multiple Django projects.
    try:
        from appsettings.mastrms import *
    except ImportError, e:
        pass

This means you can create an appsettings.mastrms module anywhere accessible to this app (using PYTHONPATH for example). The file structure would look like this:

| appsettings
|     __init__.py
|     mastrms
|         __init__.py

Then, all your overrides go in the mastrms __init__.py

Required Variable Overrides
===========================

**EMAIL_HOST** = "example - smtp.yoursite.com"
    You should replace this string with the name of your mail server.
**SERVER_EMAIL** = "example - apache@yoursite.com"
    This should be the email address which shows up as the 'From email' in emails from this webapp to users. 
**RETURN_EMAIL** = "example - noreply@yoursite.com"
    This is the return address, i.e. the email address that would be used if users replied to webapp emails.
**EMAIL_SUBJECT_PREFIX** = "DEV "
    This is a prefix for the subject line of the email, often used to differentiate emails from the live system from development system emails.
**LOGS_TO_EMAIL** = "log_email@yoursite.com"
    When a user clicks the 'Send Logs' button on the :ref:`sync client <datasync_client>`, this is the email address that gets notified. 
**KEYS_TO_EMAIL** = "key_email@yoursite.com"
    When the :ref:`sync client <datasync_client>` 'Send Keys' button is pressed, this email addres is notified.
**SECRET_KEY** = 'some random string'
    Key used for site hashing algorithms. Set this to a random string.
**DATABASES** = {a database dictionary}
    This is a python dictionary which defines how django should connect to the database

Example::
    
    {
        'default': {
            'ENGINE': 'django.db.backends.postgresql_psycopg2',
            'NAME': '<your database name here>',
            'USER': '<your database user here>',
            'PASSWORD': '<your database user password here>',
            'HOST': '<your database host here>',                      
            'PORT': '',                      
        }
    }

    #The ENGINE setting here describes a postgres database connection
    #NAME is the database name
    #USER is the database user
    #PASSWORD is the database user password
    #HOST/PORT describes your servers hostname and port.



.. _site_configuration_optional_variables:

Optional Variable Overrides
===========================

**TMP_DIRECTORY** = os.path.join(PROJECT_DIRECTORY, 'tmp')
    This must be a writeable directory where temporary files will be stored.
**DEBUG** = True
    Defines (among other things) how error messages are shown and what information is exposed. Should be False for production servers.
    
**SSL_ENABLED** = True
    Enables/Disables the SSLRedirectMiddleware. Should be False only for development, should be True on production servers.

**PERSISTENT_FILESTORE** = os.path.normpath(os.path.join(PROJECT_DIRECTORY, '..', '..', 'files'))
    An area of the filesystem for the application to write persistent data - quotes, experiment files, user uploads etc. More on the persistent filestore is covered in :ref:`Persistent Filestore <persistent_filestore>`
**MADAS_SESSION_TIMEOUT** = 1800
    Web session timeout - Defaults to 30 mins.
**CHMOD_USER** = 'apache'
    The UNIX user who should own files created by this application. If this is not the user who is running the webserver, that user should at least be in the group mentioned below, or the application will not be able to read the files it has written.
**CHMOD_GROUP** = 'maupload'
    The UNIX group who should own files created by this webapp. The user who runs the webserver should be a member of this group. This is important not onlu if CHMOD_USER is not the webserver user, but also because the :ref:`datasync <datasync_client>` process may create files not owned by the webserver user, but they will always be owned by this group.



mastrms/settings.py
# -*- coding: utf-8 -*-
"""
Django settings import defaults for development server
"""
import os
from ccg.utils.webhelpers import url

# SCRIPT_NAME isnt set when not under wsgi
if not os.environ.has_key('SCRIPT_NAME'):
    os.environ['SCRIPT_NAME']=''

SCRIPT_NAME =   os.environ['SCRIPT_NAME']
PROJECT_DIRECTORY = os.environ['PROJECT_DIRECTORY']
TMP_DIRECTORY = os.path.join(PROJECT_DIRECTORY, 'tmp')

#general site config
DEBUG = True
SITE_ID = 1
APPEND_SLASH = True
SSL_ENABLED = True

# Locals ma
TIME_ZONE = 'Australia/Perth'
LANGUAGE_CODE = 'en-us'
USE_I18N = True

##
## Django Core stuff
##
TEMPLATE_LOADERS = [
    'ccg.template.loaders.makoloader.filesystem.Loader',
    'django.template.loaders.filesystem.Loader',
    'django.template.loaders.app_directories.Loader',
]

MIDDLEWARE_CLASSES = [
    'django.middleware.common.CommonMiddleware',
    'django.contrib.sessions.middleware.SessionMiddleware',
    'django.contrib.auth.middleware.AuthenticationMiddleware',
    'django.middleware.transaction.TransactionMiddleware',
    'mastrms.utils.json_exception_handler_middleware.JSONExceptionHandlerMiddleware',
    'django.middleware.doc.XViewMiddleware',
    'ccg.middleware.ssl.SSLRedirect',
    #'ccg.middleware.StatsMiddleware.StatsMiddleware'
]

INSTALLED_APPS = [
    'django.contrib.auth',
    'django.contrib.contenttypes',
    'django.contrib.sessions',
    'django.contrib.sites',
    'django.contrib.admin',
    'django_extensions'
]

AUTHENTICATION_BACKENDS = [
 'mastrms.repository.backend.MadasBackend',
 #'ccg.auth.backends.NoAuthModelBackend',
]

#email
EMAIL_HOST = 'example - smtp.yoursite.com'                      # Address of your smtp server
SERVER_EMAIL = "example - apache@yoursite.com"                  # from address for app emails to users
RETURN_EMAIL = "example - noreply@yoursite.com"                 # reply address for app emails to users
EMAIL_SUBJECT_PREFIX = "DEV "                           # email subject prefix
# default emails in case of server tracebacks, app failures, etc
ADMINS = [
    ( 'Tech Alerts', 'example - alerts@yoursite.com' )
]
MANAGERS = ADMINS

REGISTRATION_TO_EMAIL = 'example - someadmin@yoursite.com'        #An admin who will handle site registration emails

## Ldap
AUTH_LDAP_SERVER = '<your ldap server here>'
AUTH_LDAP_BASE = '<ldap base path>'
AUTH_LDAP_GROUP_BASE = '<ldap path to group base>'
AUTH_LDAP_ADMIN_BASE = '<ldap path to admin base>'
AUTH_LDAP_USER_BASE = 'ou=NEMA,ou=People,' + AUTH_LDAP_ADMIN_BASE

AUTH_LDAP_GROUPOC = 'groupofuniquenames'
AUTH_LDAP_USEROC = 'inetorgperson'
AUTH_LDAP_MEMBERATTR = 'uniqueMember'
AUTH_LDAP_USERDN = 'ou=People'


LDAPADMINUSERNAME = '<ldap admin path>'
LDAPADMINPASSWORD = '<your ldap server password here>'
AUTH_LDAP_GROUP = 'User'
DEFAULT_GROUP = 'madas'                 # Base group for users. Needs to exist in the database.
LDAP_DONT_REQUIRE_CERT = True           # dont require HTTPS for dev ldap

#Server side directories
PERSISTENT_FILESTORE = os.path.normpath(os.path.join(PROJECT_DIRECTORY, '..', '..', 'files'))
#Ensure the persistent storage dir exits. If it doesn't, exit noisily.
assert (os.path.exists(PROJECT_DIRECTORY),
        "This application cannot start: Cannot find the project directory %s" % PROJECT_DIRECTORY)
assert (os.path.exists(PERSISTENT_FILESTORE),
        "This application cannot start: It expects a writeable directory at %s to use as a persistent filestore" % PERSISTENT_FILESTORE)
# for local development, this is set to the static serving directory. For deployment use Apache Alias
STATIC_SERVER_PATH = os.path.join(PROJECT_DIRECTORY,"static")
# a directory that will be writable by the webserver, for storing various files...
WRITABLE_DIRECTORY = os.path.join(PROJECT_DIRECTORY,"scratch")
TEMPLATE_DIRS = [
    os.path.join(PROJECT_DIRECTORY,"templates"),
]
# mako compiled templates directory
MAKO_MODULE_DIR = os.path.join(WRITABLE_DIRECTORY, "templates")
# mako module name
MAKO_MODULENAME_CALLABLE = ''

MEDIA_ROOT = os.path.join(PROJECT_DIRECTORY,"static")
MEDIA_URL = '/static/'
ADMIN_MEDIA_PREFIX = url('/static/admin-media/')

TEMPLATE_DEBUG = DEBUG
LOGIN_URL = url('/accounts/login/')
LOGOUT_URL = url('/accounts/logout/')


#session and cookies
MADAS_SESSION_TIMEOUT = 1800 #30 minute session timeout 
SESSION_COOKIE_PATH = url('/')
SESSION_SAVE_EVERY_REQUEST = True
CSRF_COOKIE_NAME = "csrftoken_madas_repoadmin"
SESSION_COOKIE_AGE = 60*60
SESSION_COOKIE_HTTPONLY = True
SESSION_COOKIE_SECURE = True

#functions to evaluate for status checking
from status_checks import *
STATUS_CHECKS = [check_default]

# memcache server list
MEMCACHE_SERVERS = ['yourmemcacheserver.yourdomain:11211','anothermemcacheserver.yourdomain:11211']
MEMCACHE_KEYSPACE = "mastrms-dev"


#APPLICATION SPECIFIC SETTINGS
ROOT_URLCONF = 'mastrms.urls'
SITE_NAME = 'madas'
SECRET_KEY = 'qj#tl@9@7((%^)$i#iyw0gcfzf&#a*pobgb8yr#1%65+*6!@g$'
EMAIL_APP_NAME = "MastrMS "
REPO_FILES_ROOT = PERSISTENT_FILESTORE
#quote files root should be within the repo files root
QUOTE_FILES_ROOT = os.path.join(REPO_FILES_ROOT, 'quotes')

INSTALLED_APPS.extend([
    'django_extensions',
    'mastrms.mdatasync_server',
    'mastrms.login',
    'mastrms.quote',
    'mastrms.admin',
    'mastrms.repository',
    'mastrms.users',
    'south'])

DATABASES = {
    'default': {
        'ENGINE': 'django.db.backends.postgresql_psycopg2',
        'NAME': '<your database name here>',
        'USER': '<your database user here>',
        'PASSWORD': '<your database user password here>',
        'HOST': '<your database host here>',                      
        'PORT': '',                      
    }
}

CHMOD_USER = 'apache'
CHMOD_GROUP = 'maupload'

#Datasync email settings:
LOGS_TO_EMAIL = "log_email@yoursite.com" #email address to receive datasync client log notifications
KEYS_TO_EMAIL = "key_email@yoursite.com" #email address to receive datasync key upload notifications


##
## LOGGING
##

LOG_DIRECTORY = os.path.join(PROJECT_DIRECTORY, "logs")
try:
    if not os.path.exists(LOG_DIRECTORY):
        os.mkdir(LOG_DIRECTORY)
except:
    pass
os.path.exists(LOG_DIRECTORY), "No logs directory, please create one: %s" % LOG_DIRECTORY
INSTALL_NAME = PROJECT_DIRECTORY.split('/')[-2]

LOGGING = {
    'version': 1,
    'disable_existing_loggers': True,
    'formatters': {
        'verbose': {
            'format': 'MASTRMS [%(name)s:' + INSTALL_NAME + ':%(levelname)s:%(asctime)s:%(filename)s:%(lineno)s:%(funcName)s] %(message)s'
        },
        'db': {
            'format': 'MASTRMS [%(name)s:' + INSTALL_NAME + ':%(duration)s:%(sql)s:%(params)s] %(message)s'
        },
        'simple': {
            'format': 'MASTRMS ' + INSTALL_NAME + ' %(levelname)s %(message)s'
        },
    },
    'filters': {
    },
    'handlers': {
        'null': {
            'level':'DEBUG',
            'class':'django.utils.log.NullHandler',
        },
        'console':{
            'level':'DEBUG',
            'class':'logging.StreamHandler',
            'formatter': 'verbose'
        },
        'errorfile':{
            'level':'ERROR',
            'class':'logging.handlers.TimedRotatingFileHandler',
            'filename': os.path.join(LOG_DIRECTORY, 'error.log'),
            'when':'midnight',
            'formatter': 'verbose'
        },
        'madasfile':{
            'class':'logging.handlers.TimedRotatingFileHandler',
            'filename': os.path.join(LOG_DIRECTORY, 'mastrms.log'),
            'when':'midnight',
            'formatter': 'verbose'
        },
        'mdatasyncfile':{
            'class':'logging.handlers.TimedRotatingFileHandler',
            'filename': os.path.join(LOG_DIRECTORY, 'mdatasync_server.log'),
            'when':'midnight',
            'formatter': 'verbose'
        },
        'db_logfile':{
            'level':'DEBUG',
            'class':'logging.handlers.TimedRotatingFileHandler',
            'filename': os.path.join(LOG_DIRECTORY, 'mastrms_db.log'),
            'when':'midnight',
            'formatter': 'db'
        },
        'mail_admins': {
            'level': 'ERROR',
            'class': 'django.utils.log.AdminEmailHandler',
            'formatter':'verbose',
            'include_html':True
        }
    },
    'root': {
            'handlers':['console', 'errorfile', 'mail_admins'],
            'level':'ERROR',
    },
    'loggers': {
        'django': {
            'handlers':['null'],
            'propagate': False,
            'level':'INFO',
        },
        'madas_log': {
            'handlers': ['madasfile'],
            'level': 'INFO',
            'propagate': False,
        },
        'mdatasync_server_log': {
            'handlers': ['mdatasyncfile'],
            'level': 'INFO',
            'propagate': True,
        }
    }
}


# Override defaults with your local instance settings.
# They will be loaded from appsettings.<projectname>, which can exist anywhere
# in the instance's pythonpath. This is a CCG convention designed to support
# global shared settings among multiple Django projects.
try:
    from appsettings.mastrms import *
except ImportError, e:
    pass

#STATS_MIDDLEWARE_LOGGER = 'mastrms_log'
